lib/module.js

```diff
@@ -434,7 +435,34 @@ Module._load = function(request, parent, isMain) {
     debug('Module._load REQUEST %s parent: %s', request, parent.id);
   }
 
-  var filename = Module._resolveFilename(request, parent, isMain);
+  var filename = null;
+
+  if (isMain) {
+    let err;
+    try {
+      filename = Module._resolveFilename(request, parent, isMain);
+    } catch (e) {
+      // try to keep stack
+      e.stack;
+      err = e;
+    }
+    if (filename === null || /\.mjs$/.test(filename)) {
+      try {
+        ESM.loader.import(request).catch((e) => {
+          console.error(e);
+          process.exit(1);
+        });
+        return;
```

Comment:
@TimothyGu caching works, but it seems there was a regression looking it over. It should not load modules from scratch if they are already loaded in require.cache when they are imported. This is not visible due to it being only in runMain though